### PR TITLE
private query dependencies

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by nikomatsakis** [comment]
[Zulip thread for discussion is here](https://salsa.zulipchat.com/#narrow/stream/198105-rfcs.2Fprivate-query-dependencies)

File: RFC0003-Query-Dependencies.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    fn get_symbol_by_name(&self, name: String) -> Symbol;
+}
+```
+
+The value of `requires` is a `+` separated list of traits, like the one you would
+write in a `where` clause. These traits are available when implementing the query:
+
+```rust
+fn get_symbol_by_name(
+    db: &impl SymbolsDatabase + SyntaxDatabase + EnvDatabase,
```
Comment (kestred) [inline RFC0003-Query-Dependencies.md]: AFAIK, the compiler requires parens here `&(impl A + B + C)`,
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7d81641234462bb2a2f6daa55afadd98

File: RFC0003-Query-Dependencies.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+```
+
+Note that, while the RFC does not propose to add per-query dependencies, query
+implementation can voluntarily specify only a subset of traits from `requires`
+clause:
+
+```rust
+fn get_symbol_by_name(
+    // Purposefully don't depend on EnvDatabase
+    db: &impl SymbolsDatabase + SyntaxDatabase,
```
Comment (nikomatsakis) [inline RFC0003-Query-Dependencies.md]:
```suggestion
    db: &(impl SymbolsDatabase + SyntaxDatabase),
```

File: RFC0003-Query-Dependencies.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    fn get_symbol_by_name(&self, name: String) -> Symbol;
+}
+```
+
+The value of `requires` is a `+` separated list of traits, like the one you would
+write in a `where` clause. These traits are available when implementing the query:
+
+```rust
+fn get_symbol_by_name(
+    db: &impl SymbolsDatabase + SyntaxDatabase + EnvDatabase,
```
Comment (nikomatsakis) [inline RFC0003-Query-Dependencies.md]:
```suggestion
    db: &(impl SymbolsDatabase + SyntaxDatabase + EnvDatabase),
```

File: RFC0003-Query-Dependencies.md (hunk 1 above)
Comment (matklad) [reply RFC0003-Query-Dependencies.md]: Thanks, fixed!

**Comment by matklad** [comment]
@nikomatsakis this is implemented and released, could you merge the PR?

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**